addon/utils/should-add-token.js

```diff
@@ -1,36 +1,30 @@
-export default function shouldAddToken (url, portalId = '', portalHost) {
-  // if this is a hosted service that belongs to the user's organization
-  // if this is a service with stored credentials
-  const id = portalId.toLowerCase();
-  if (portalHost) {
-    return hostsMatch(portalHost, url);
-  } else {
-    return (
-      !!url.toLowerCase().match('arcgis.com/arcgis/rest/services') ||
-      !!url.toLowerCase().match(`arcgis.com/${id}/arcgis/rest`) ||
-      /utility\w+\.arcgis\.com\/sharing\/servers\/\w{32}\/rest\/services\/Hosted/i.test(url)
-    );
+export default function shouldAddToken (url, serverInfo, portalInfo) {
+  let hostsDidMatch = false;
+  if (portalInfo.portalHostname) {
+    // make sure either the server and the portal are on the same domain
+    hostsDidMatch = hostsMatch(portalInfo.portalHostname, serverInfo.owningSystemUrl);
   }
+
+  // or it's in authorizedCrossOriginDomains
+  const domain = stripToDomain(url);
+  const isAuthorizedUrl = portalInfo.authorizedCrossOriginDomains.includes(domain);
```

Comment:
should probably be defensive here

Reply:
to support AGO, I think we need to skip this check if both the portal url and the service url are on *.arcgis.com

Reply:
@dmfenton - that's the exact case I'm dealing with. I've got a situation where `portalHostname` is "qaext.arcgis.com" and the service url is "https://services2.arcgis.com/zNjnZafDYCAJAbN0/arcgis/rest/services/Street_ROW_Trees/FeatureServer". If we send the token in that case we get "invalid token".

Reply:
right. So the check becomes `if arcgisDomain && hostsMatch => send token`